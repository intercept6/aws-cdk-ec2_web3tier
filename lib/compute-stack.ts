import cdk = require('@aws-cdk/core');
import {AutoScalingGroup} from "@aws-cdk/aws-autoscaling";
import {
    AmazonLinuxImage,
    InstanceClass,
    InstanceSize,
    InstanceType,
    ISecurityGroup,
    IVpc,
    SubnetType
} from "@aws-cdk/aws-ec2";
import {ApplicationLoadBalancer} from "@aws-cdk/aws-elasticloadbalancingv2";
import {Role} from "@aws-cdk/aws-iam";

interface ComputeStackProps extends cdk.StackProps {
    vpc: IVpc;
    instanceRole: Role;
    appSg: ISecurityGroup;
}

export class ComputeStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);


        const prj: string = this.node.tryGetContext('prj');
        const stage: string = this.node.tryGetContext('stage');
        // const params: any = this.node.tryGetContext(stage);

        const asg = new AutoScalingGroup(this, 'ASG', {
            vpc: props.vpc,
            instanceType: InstanceType.of(InstanceClass.BURSTABLE2, InstanceSize.MICRO),
            machineImage: new AmazonLinuxImage(),
            allowAllOutbound: true,
            role: props.instanceRole
        });
        asg.addUserData(
            "yum -y update",
            "yum -y install nginx",
            "systemctl enable nginx",
            "systemctl start nginx"
        );

        const alb = new ApplicationLoadBalancer(this, 'Alb', {
            vpc: props.vpc,
            vpcSubnets: {subnetType: SubnetType.PUBLIC},
            internetFacing: true,
            loadBalancerName: `${prj}-${stage}-alb`
        });

        const listener = alb.addListener('Listener', {
            port: 80,
        });
        listener.addTargets('Target', {
            targets: [asg],
            port: 80
        });

    }
}