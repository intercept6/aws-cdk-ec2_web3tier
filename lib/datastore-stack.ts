// @aws-cdk/aws-rds is a developer preview (public beta) module.
// Releases might lack important features and might have future breaking changes.

import {
  Construct,
  RemovalPolicy,
  Stack,
  StackProps,
  Tag
} from "@aws-cdk/core";
import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  ISecurityGroup,
  IVpc,
  SubnetType
} from "@aws-cdk/aws-ec2";
import { DatabaseCluster, DatabaseClusterEngine } from "@aws-cdk/aws-rds";

interface DataStoreStackProps extends StackProps {
  vpc: IVpc;
  dbSg: ISecurityGroup;
}

export class DataStoreStack extends Stack {
  constructor(scope: Construct, id: string, props: DataStoreStackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: any = this.node.tryGetContext(stage);

    const cluster = new DatabaseCluster(this, "DBCluster", {
      engine: DatabaseClusterEngine.AURORA,
      masterUser: {
        username: "admin"
      },
      removalPolicy: RemovalPolicy.DESTROY,
      instanceProps: {
        instanceType: InstanceType.of(
          InstanceClass.MEMORY5,
          InstanceSize.LARGE
        ),
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED
        },
        vpc: props.vpc,
        securityGroup: props.dbSg
      }
    });
    cluster.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
