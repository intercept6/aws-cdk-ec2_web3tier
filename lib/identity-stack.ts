import { Construct, Stack, StackProps } from "@aws-cdk/core";
import { ManagedPolicy, Role, ServicePrincipal } from "@aws-cdk/aws-iam";

export class IdentityStack extends Stack {
  public readonly instanceRole: Role;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: object = this.node.tryGetContext(stage);

    this.instanceRole = new Role(this, "IamRole", {
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      roleName: `${prj}-${stage}-app`,
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonEC2RoleforSSM"
        )
      ]
    });
  }
}
