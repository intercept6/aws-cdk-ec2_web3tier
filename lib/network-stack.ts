import { Construct, Stack, StackProps, Tag } from "@aws-cdk/core";
import { Port, SecurityGroup, SubnetType, Vpc } from "@aws-cdk/aws-ec2";

export class NetworkStack extends Stack {
    public readonly vpc: Vpc;
    public readonly appSg: SecurityGroup;
    public readonly dbSg: SecurityGroup;

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const prj: string = this.node.tryGetContext("prj");
        const stage: string = this.node.tryGetContext("stage");
        const params: any = this.node.tryGetContext(stage);

    this.vpc = new Vpc(this, "VPC", {
      enableDnsHostnames: true,
      enableDnsSupport: true,
      natGateways: params.nat_gateways,
      maxAzs: 2,
      subnetConfiguration: [
        { name: "Public", subnetType: SubnetType.PUBLIC, cidrMask: 24 },
        { name: "Private", subnetType: SubnetType.PRIVATE, cidrMask: 24 },
        { name: "Isolated", subnetType: SubnetType.ISOLATED, cidrMask: 24 }
      ]
    });
    this.vpc.node.applyAspect(new Tag("Name", `${prj}-${stage}-vpc`));

    this.appSg = new SecurityGroup(this, "AppSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-app`,
      description: `${prj}-${stage}-app`
    });
    this.appSg.addIngressRule(
      this.appSg,
      Port.tcp(80),
      "allow http access from alb"
    );
    this.appSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-app`));

    this.dbSg = new SecurityGroup(this, "DbSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-db`,
      description: `${prj}-${stage}-db`
    });
    this.dbSg.addIngressRule(
      this.appSg,
      Port.tcp(3306),
      "allow db access from app server"
    );
    this.dbSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
