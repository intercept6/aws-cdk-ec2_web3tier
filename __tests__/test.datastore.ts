import { App, Stack } from "@aws-cdk/core";
import { SynthUtils } from "@aws-cdk/assert";
import { DataStoreStack } from "../lib/datastore-stack";
import { SecurityGroup, SubnetType, Vpc } from "@aws-cdk/aws-ec2";

describe("datastore", () => {
  test("default", () => {
    const app = new App();
    app.node.setContext("prj", "ssprj");
    app.node.setContext("stage", "ssstage");
    app.node.setContext("ssstage", {
      nat_gateways: 1
    });
    const givenStack = new Stack(app, "GivenStack");
    const vpc = new Vpc(givenStack, "TestVPC", {
      subnetConfiguration: [
        { name: "Isolated", subnetType: SubnetType.ISOLATED }
      ]
    });
    const sg = new SecurityGroup(givenStack, "TestSecurityGroup", {
      vpc: vpc
    });
    const stack = new DataStoreStack(app, "TestDataStoreStack", {
      vpc: vpc,
      dbSg: sg
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
