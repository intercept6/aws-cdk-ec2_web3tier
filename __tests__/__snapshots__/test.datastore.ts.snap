// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datastore default 1`] = `
Object {
  "Resources": Object {
    "DBCluster15AF587F": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBSubnetGroupName": Object {
          "Ref": "DBClusterSubnets1338ED0E",
        },
        "Engine": "aurora",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DBClusterSecretCDFF091A",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "DBClusterSecretCDFF091A",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-db",
          },
        ],
        "VpcSecurityGroupIds": Array [
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputFnGetAttTestSecurityGroup880B57C0GroupIdE8B1161D",
          },
        ],
      },
      "Type": "AWS::RDS::DBCluster",
      "UpdateReplacePolicy": "Delete",
    },
    "DBClusterInstance160268105": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterIdentifier": Object {
          "Ref": "DBCluster15AF587F",
        },
        "DBInstanceClass": "db.r5.large",
        "DBSubnetGroupName": Object {
          "Ref": "DBClusterSubnets1338ED0E",
        },
        "Engine": "aurora",
        "PubliclyAccessible": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-db",
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "DBClusterInstance2E3B28A27": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterIdentifier": Object {
          "Ref": "DBCluster15AF587F",
        },
        "DBInstanceClass": "db.r5.large",
        "DBSubnetGroupName": Object {
          "Ref": "DBClusterSubnets1338ED0E",
        },
        "Engine": "aurora",
        "PubliclyAccessible": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-db",
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "DBClusterSecretAttachedSecret7283BE31": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "DBClusterSecretCDFF091A",
        },
        "TargetId": Object {
          "Ref": "DBCluster15AF587F",
        },
        "TargetType": "AWS::RDS::DBCluster",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "DBClusterSecretCDFF091A": Object {
      "Properties": Object {
        "GenerateSecretString": Object {
          "ExcludeCharacters": "\\"@/\\\\",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{\\"username\\":\\"admin\\"}",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-db",
          },
        ],
      },
      "Type": "AWS::SecretsManager::Secret",
    },
    "DBClusterSubnets1338ED0E": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnets for DBCluster database",
        "SubnetIds": Array [
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCIsolatedSubnet1SubnetB4ED5A5C1971312B",
          },
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCIsolatedSubnet2Subnet84140C2BE236F491",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-db",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
  },
}
`;
